File: qt/spatialaudio/gen_qaudioengine.cpp
// +build ignore

#include <QAudioDevice>
#include <QAudioEngine>
#include <QMetaObject>
#include <QObject>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qaudioengine.h>
#include "gen_qaudioengine.h"
class MiqtVirtualQAudioEngine : public virtual QAudioEngine {
public:
MiqtVirtualQAudioEngine(): QAudioEngine() {};
MiqtVirtualQAudioEngine(QObject* parent): QAudioEngine(parent) {};
MiqtVirtualQAudioEngine(int sampleRate): QAudioEngine(sampleRate) {};
MiqtVirtualQAudioEngine(int sampleRate, QObject* parent): QAudioEngine(sampleRate, parent) {};

virtual ~MiqtVirtualQAudioEngine() = default;
// cgo.Handle value for overwritten implementation
	intptr_t handle__MetaObject = 0;
// Subclass to allow providing a Go implementation
	virtual const QMetaObject* metaObject() const override {
if (handle__MetaObject == 0) {
return QAudioEngine::metaObject();
}
QMetaObject* callback_return_value = miqt_exec_callback_QAudioEngine_MetaObject(const_cast<MiqtVirtualQAudioEngine*>(this), handle__MetaObject);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	QMetaObject* virtualbase_MetaObject() const {

		return (QMetaObject*) QAudioEngine::metaObject();

	}
// cgo.Handle value for overwritten implementation
	intptr_t handle__Metacast = 0;
// Subclass to allow providing a Go implementation
	virtual void* qt_metacast(const char* param1) override {
if (handle__Metacast == 0) {
return QAudioEngine::qt_metacast(param1);
}
const char* sigval1 = (const char*) param1;

		void* callback_return_value = miqt_exec_callback_QAudioEngine_Metacast(this, handle__Metacast, sigval1);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	void* virtualbase_Metacast(const char* param1) {

		return QAudioEngine::qt_metacast(param1);

	}
};
QAudioEngine* QAudioEngine_new() {
return new MiqtVirtualQAudioEngine();
}
QAudioEngine* QAudioEngine_new2(QObject* parent) {
return new MiqtVirtualQAudioEngine(parent);
}
QAudioEngine* QAudioEngine_new3(int sampleRate) {
return new MiqtVirtualQAudioEngine(static_cast<int>(sampleRate));
}
QAudioEngine* QAudioEngine_new4(int sampleRate, QObject* parent) {
return new MiqtVirtualQAudioEngine(static_cast<int>(sampleRate), parent);
}
void QAudioEngine_virtbase(QAudioEngine* src
, QObject** outptr_QObject
) {
*outptr_QObject = static_cast<QObject*>(src);
}
QMetaObject* QAudioEngine_MetaObject(const QAudioEngine* self) {
	return (QMetaObject*) self->metaObject();
}
void* QAudioEngine_Metacast(QAudioEngine* self, const char* param1) {
	return self->qt_metacast(param1);
}
struct miqt_string QAudioEngine_Tr(const char* s) {
	QString _ret = QAudioEngine::tr(s);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
void QAudioEngine_SetOutputMode(QAudioEngine* self, OutputMode mode) {
	self->setOutputMode(mode);
}
OutputMode QAudioEngine_OutputMode(const QAudioEngine* self) {
	return self->outputMode();
}
int QAudioEngine_SampleRate(const QAudioEngine* self) {
	return self->sampleRate();
}
void QAudioEngine_SetOutputDevice(QAudioEngine* self, QAudioDevice* device) {
	self->setOutputDevice(*device);
}
QAudioDevice* QAudioEngine_OutputDevice(const QAudioEngine* self) {
	return new QAudioDevice(self->outputDevice());
}
void QAudioEngine_SetMasterVolume(QAudioEngine* self, float volume) {
	self->setMasterVolume(static_cast<float>(volume));
}
float QAudioEngine_MasterVolume(const QAudioEngine* self) {
	return self->masterVolume();
}
void QAudioEngine_SetPaused(QAudioEngine* self, bool paused) {
	self->setPaused(paused);
}
bool QAudioEngine_Paused(const QAudioEngine* self) {
	return self->paused();
}
void QAudioEngine_SetRoomEffectsEnabled(QAudioEngine* self, bool enabled) {
	self->setRoomEffectsEnabled(enabled);
}
bool QAudioEngine_RoomEffectsEnabled(const QAudioEngine* self) {
	return self->roomEffectsEnabled();
}
void QAudioEngine_SetDistanceScale(QAudioEngine* self, float scale) {
	self->setDistanceScale(static_cast<float>(scale));
}
float QAudioEngine_DistanceScale(const QAudioEngine* self) {
	return self->distanceScale();
}
void QAudioEngine_OutputModeChanged(QAudioEngine* self) {
	self->outputModeChanged();
}
void QAudioEngine_connect_OutputModeChanged(QAudioEngine* self, intptr_t slot) {
	MiqtVirtualQAudioEngine::connect(self, static_cast<void (QAudioEngine::*)()>(&QAudioEngine::outputModeChanged), self, [=]() {
		miqt_exec_callback_QAudioEngine_OutputModeChanged(slot);
	});
}
void QAudioEngine_OutputDeviceChanged(QAudioEngine* self) {
	self->outputDeviceChanged();
}
void QAudioEngine_connect_OutputDeviceChanged(QAudioEngine* self, intptr_t slot) {
	MiqtVirtualQAudioEngine::connect(self, static_cast<void (QAudioEngine::*)()>(&QAudioEngine::outputDeviceChanged), self, [=]() {
		miqt_exec_callback_QAudioEngine_OutputDeviceChanged(slot);
	});
}
void QAudioEngine_MasterVolumeChanged(QAudioEngine* self) {
	self->masterVolumeChanged();
}
void QAudioEngine_connect_MasterVolumeChanged(QAudioEngine* self, intptr_t slot) {
	MiqtVirtualQAudioEngine::connect(self, static_cast<void (QAudioEngine::*)()>(&QAudioEngine::masterVolumeChanged), self, [=]() {
		miqt_exec_callback_QAudioEngine_MasterVolumeChanged(slot);
	});
}
void QAudioEngine_PausedChanged(QAudioEngine* self) {
	self->pausedChanged();
}
void QAudioEngine_connect_PausedChanged(QAudioEngine* self, intptr_t slot) {
	MiqtVirtualQAudioEngine::connect(self, static_cast<void (QAudioEngine::*)()>(&QAudioEngine::pausedChanged), self, [=]() {
		miqt_exec_callback_QAudioEngine_PausedChanged(slot);
	});
}
void QAudioEngine_DistanceScaleChanged(QAudioEngine* self) {
	self->distanceScaleChanged();
}
void QAudioEngine_connect_DistanceScaleChanged(QAudioEngine* self, intptr_t slot) {
	MiqtVirtualQAudioEngine::connect(self, static_cast<void (QAudioEngine::*)()>(&QAudioEngine::distanceScaleChanged), self, [=]() {
		miqt_exec_callback_QAudioEngine_DistanceScaleChanged(slot);
	});
}
void QAudioEngine_Start(QAudioEngine* self) {
	self->start();
}
void QAudioEngine_Stop(QAudioEngine* self) {
	self->stop();
}
void QAudioEngine_Pause(QAudioEngine* self) {
	self->pause();
}
void QAudioEngine_Resume(QAudioEngine* self) {
	self->resume();
}
struct miqt_string QAudioEngine_Tr2(const char* s, const char* c) {
	QString _ret = QAudioEngine::tr(s, c);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
struct miqt_string QAudioEngine_Tr3(const char* s, const char* c, int n) {
	QString _ret = QAudioEngine::tr(s, c, static_cast<int>(n));
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
void QAudioEngine_override_virtual_MetaObject(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQAudioEngine*>( (QAudioEngine*)(self) )->handle__MetaObject = slot;
}
QMetaObject* QAudioEngine_virtualbase_MetaObject(const void* self) {
	return ( (const MiqtVirtualQAudioEngine*)(self) )->virtualbase_MetaObject();
}
void QAudioEngine_override_virtual_Metacast(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQAudioEngine*>( (QAudioEngine*)(self) )->handle__Metacast = slot;
}
void* QAudioEngine_virtualbase_Metacast(void* self, const char* param1) {
	return ( (MiqtVirtualQAudioEngine*)(self) )->virtualbase_Metacast(param1);
}
void QAudioEngine_Delete(QAudioEngine* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<MiqtVirtualQAudioEngine*>( self );
	} else {
		delete self;
	}
}

File: qt/spatialaudio/gen_qaudioengine.go
package spatialaudio

import (
	"github.com/mappu/miqt/qt"
	"unsafe"
)

type QAudioEngine__OutputMode int

const (
	QAudioEngine__Surround  QAudioEngine__OutputMode = 0
	QAudioEngine__Stereo    QAudioEngine__OutputMode = 1
	QAudioEngine__Headphone QAudioEngine__OutputMode = 2
)

type QAudioEngine struct {
	h          uintptr
	isSubclass bool
}

// NewQAudioEngine constructs a new QAudioEngine object.
func NewQAudioEngine() *QAudioEngine {

	ret := newQAudioEngine(QAudioEngine_new())
	ret.isSubclass = true
	return ret
}

// NewQAudioEngine2 constructs a new QAudioEngine object.
func NewQAudioEngine2(parent *qt.QObject) *QAudioEngine {

	ret := newQAudioEngine(QAudioEngine_new2((*QObject)(parent.UnsafePointer())))
	ret.isSubclass = true
	return ret
}

// NewQAudioEngine3 constructs a new QAudioEngine object.
func NewQAudioEngine3(sampleRate int) *QAudioEngine {

	ret := newQAudioEngine(QAudioEngine_new3((int)(sampleRate)))
	ret.isSubclass = true
	return ret
}

// NewQAudioEngine4 constructs a new QAudioEngine object.
func NewQAudioEngine4(sampleRate int, parent *qt.QObject) *QAudioEngine {

	ret := newQAudioEngine(QAudioEngine_new4((int)(sampleRate), (*QObject)(parent.UnsafePointer())))
	ret.isSubclass = true
	return ret
}

func (this *QAudioEngine) MetaObject() *qt.QMetaObject {
	return qt.UnsafeNewQMetaObject(unsafe.Pointer(QAudioEngine_MetaObject(this.h)))
}

func (this *QAudioEngine) Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))
	return (unsafe.Pointer)(QAudioEngine_Metacast(this.h, param1_Cstring))
}

func QAudioEngine_Tr(s string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	var _ms struct_miqt_string = QAudioEngine_Tr(s_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QAudioEngine) SetOutputMode(mode OutputMode) {
	QAudioEngine_SetOutputMode(this.h, mode)
}

func (this *QAudioEngine) OutputMode() OutputMode {
	xxxxxxxxx
}

func (this *QAudioEngine) SampleRate() int {
	return (int)(QAudioEngine_SampleRate(this.h))
}

func (this *QAudioEngine) SetOutputDevice(device *QAudioDevice) {
	QAudioEngine_SetOutputDevice(this.h, device)
}

func (this *QAudioEngine) OutputDevice() QAudioDevice {
	xxxxxxxxx
}

func (this *QAudioEngine) SetMasterVolume(volume float32) {
	QAudioEngine_SetMasterVolume(this.h, (float)(volume))
}

func (this *QAudioEngine) MasterVolume() float32 {
	return (float32)(QAudioEngine_MasterVolume(this.h))
}

func (this *QAudioEngine) SetPaused(paused bool) {
	QAudioEngine_SetPaused(this.h, (bool)(paused))
}

func (this *QAudioEngine) Paused() bool {
	return (bool)(QAudioEngine_Paused(this.h))
}

func (this *QAudioEngine) SetRoomEffectsEnabled(enabled bool) {
	QAudioEngine_SetRoomEffectsEnabled(this.h, (bool)(enabled))
}

func (this *QAudioEngine) RoomEffectsEnabled() bool {
	return (bool)(QAudioEngine_RoomEffectsEnabled(this.h))
}

func (this *QAudioEngine) SetDistanceScale(scale float32) {
	QAudioEngine_SetDistanceScale(this.h, (float)(scale))
}

func (this *QAudioEngine) DistanceScale() float32 {
	return (float32)(QAudioEngine_DistanceScale(this.h))
}

func (this *QAudioEngine) OutputModeChanged() {
	QAudioEngine_OutputModeChanged(this.h)
}
func (this *QAudioEngine) OnOutputModeChanged(slot func()) {
	QAudioEngine_connect_OutputModeChanged(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QAudioEngine_OutputModeChanged
func miqt_exec_callback_QAudioEngine_OutputModeChanged(cb intptr_t) {
	gofunc, ok := cgo.Handle(cb).Value().(func())
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	gofunc()
}

func (this *QAudioEngine) OutputDeviceChanged() {
	QAudioEngine_OutputDeviceChanged(this.h)
}
func (this *QAudioEngine) OnOutputDeviceChanged(slot func()) {
	QAudioEngine_connect_OutputDeviceChanged(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QAudioEngine_OutputDeviceChanged
func miqt_exec_callback_QAudioEngine_OutputDeviceChanged(cb intptr_t) {
	gofunc, ok := cgo.Handle(cb).Value().(func())
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	gofunc()
}

func (this *QAudioEngine) MasterVolumeChanged() {
	QAudioEngine_MasterVolumeChanged(this.h)
}
func (this *QAudioEngine) OnMasterVolumeChanged(slot func()) {
	QAudioEngine_connect_MasterVolumeChanged(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QAudioEngine_MasterVolumeChanged
func miqt_exec_callback_QAudioEngine_MasterVolumeChanged(cb intptr_t) {
	gofunc, ok := cgo.Handle(cb).Value().(func())
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	gofunc()
}

func (this *QAudioEngine) PausedChanged() {
	QAudioEngine_PausedChanged(this.h)
}
func (this *QAudioEngine) OnPausedChanged(slot func()) {
	QAudioEngine_connect_PausedChanged(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QAudioEngine_PausedChanged
func miqt_exec_callback_QAudioEngine_PausedChanged(cb intptr_t) {
	gofunc, ok := cgo.Handle(cb).Value().(func())
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	gofunc()
}

func (this *QAudioEngine) DistanceScaleChanged() {
	QAudioEngine_DistanceScaleChanged(this.h)
}
func (this *QAudioEngine) OnDistanceScaleChanged(slot func()) {
	QAudioEngine_connect_DistanceScaleChanged(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QAudioEngine_DistanceScaleChanged
func miqt_exec_callback_QAudioEngine_DistanceScaleChanged(cb intptr_t) {
	gofunc, ok := cgo.Handle(cb).Value().(func())
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	gofunc()
}

func (this *QAudioEngine) Start() {
	QAudioEngine_Start(this.h)
}

func (this *QAudioEngine) Stop() {
	QAudioEngine_Stop(this.h)
}

func (this *QAudioEngine) Pause() {
	QAudioEngine_Pause(this.h)
}

func (this *QAudioEngine) Resume() {
	QAudioEngine_Resume(this.h)
}

func QAudioEngine_Tr2(s string, c string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QAudioEngine_Tr2(s_Cstring, c_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func QAudioEngine_Tr3(s string, c string, n int) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QAudioEngine_Tr3(s_Cstring, c_Cstring, (int)(n))
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QAudioEngine) callVirtualBase_Event(event *qt.QEvent) bool {

	return (bool)(QAudioEngine_virtualbase_Event(unsafe.Pointer(this.h), (*QEvent)(event.UnsafePointer())))

}
func (this *QAudioEngine) OnEvent(slot func(super func(event *qt.QEvent) bool, event *qt.QEvent) bool) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QAudioEngine_override_virtual_Event(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QAudioEngine_Event
func miqt_exec_callback_QAudioEngine_Event(self QAudioEngine, cb intptr_t, event *QEvent) bool {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(event *qt.QEvent) bool, event *qt.QEvent) bool)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	slotval1 := qt.UnsafeNewQEvent(unsafe.Pointer(event))

	virtualReturn := gofunc((&QAudioEngine{h: self}).callVirtualBase_Event, slotval1)

	return (bool)(virtualReturn)

}

func (this *QAudioEngine) callVirtualBase_EventFilter(watched *qt.QObject, event *qt.QEvent) bool {

	return (bool)(QAudioEngine_virtualbase_EventFilter(unsafe.Pointer(this.h), (*QObject)(watched.UnsafePointer()), (*QEvent)(event.UnsafePointer())))

}
func (this *QAudioEngine) OnEventFilter(slot func(super func(watched *qt.QObject, event *qt.QEvent) bool, watched *qt.QObject, event *qt.QEvent) bool) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QAudioEngine_override_virtual_EventFilter(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QAudioEngine_EventFilter
func miqt_exec_callback_QAudioEngine_EventFilter(self QAudioEngine, cb intptr_t, watched *QObject, event *QEvent) bool {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(watched *qt.QObject, event *qt.QEvent) bool, watched *qt.QObject, event *qt.QEvent) bool)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	slotval1 := qt.UnsafeNewQObject(unsafe.Pointer(watched))

	slotval2 := qt.UnsafeNewQEvent(unsafe.Pointer(event))

	virtualReturn := gofunc((&QAudioEngine{h: self}).callVirtualBase_EventFilter, slotval1, slotval2)

	return (bool)(virtualReturn)

}

func (this *QAudioEngine) callVirtualBase_TimerEvent(event *qt.QTimerEvent) {

	QAudioEngine_virtualbase_TimerEvent(unsafe.Pointer(this.h), (*QTimerEvent)(event.UnsafePointer()))

}
func (this *QAudioEngine) OnTimerEvent(slot func(super func(event *qt.QTimerEvent), event *qt.QTimerEvent)) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QAudioEngine_override_virtual_TimerEvent(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QAudioEngine_TimerEvent
func miqt_exec_callback_QAudioEngine_TimerEvent(self QAudioEngine, cb intptr_t, event *QTimerEvent) {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(event *qt.QTimerEvent), event *qt.QTimerEvent))
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	slotval1 := qt.UnsafeNewQTimerEvent(unsafe.Pointer(event))

	gofunc((&QAudioEngine{h: self}).callVirtualBase_TimerEvent, slotval1)

}

func (this *QAudioEngine) callVirtualBase_ChildEvent(event *qt.QChildEvent) {

	QAudioEngine_virtualbase_ChildEvent(unsafe.Pointer(this.h), (*QChildEvent)(event.UnsafePointer()))

}
func (this *QAudioEngine) OnChildEvent(slot func(super func(event *qt.QChildEvent), event *qt.QChildEvent)) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QAudioEngine_override_virtual_ChildEvent(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QAudioEngine_ChildEvent
func miqt_exec_callback_QAudioEngine_ChildEvent(self QAudioEngine, cb intptr_t, event *QChildEvent) {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(event *qt.QChildEvent), event *qt.QChildEvent))
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	slotval1 := qt.UnsafeNewQChildEvent(unsafe.Pointer(event))

	gofunc((&QAudioEngine{h: self}).callVirtualBase_ChildEvent, slotval1)

}

func (this *QAudioEngine) callVirtualBase_CustomEvent(event *qt.QEvent) {

	QAudioEngine_virtualbase_CustomEvent(unsafe.Pointer(this.h), (*QEvent)(event.UnsafePointer()))

}
func (this *QAudioEngine) OnCustomEvent(slot func(super func(event *qt.QEvent), event *qt.QEvent)) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QAudioEngine_override_virtual_CustomEvent(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QAudioEngine_CustomEvent
func miqt_exec_callback_QAudioEngine_CustomEvent(self QAudioEngine, cb intptr_t, event *QEvent) {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(event *qt.QEvent), event *qt.QEvent))
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	slotval1 := qt.UnsafeNewQEvent(unsafe.Pointer(event))

	gofunc((&QAudioEngine{h: self}).callVirtualBase_CustomEvent, slotval1)

}

func (this *QAudioEngine) callVirtualBase_ConnectNotify(signal *qt.QMetaMethod) {

	QAudioEngine_virtualbase_ConnectNotify(unsafe.Pointer(this.h), (*QMetaMethod)(signal.UnsafePointer()))

}
func (this *QAudioEngine) OnConnectNotify(slot func(super func(signal *qt.QMetaMethod), signal *qt.QMetaMethod)) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QAudioEngine_override_virtual_ConnectNotify(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QAudioEngine_ConnectNotify
func miqt_exec_callback_QAudioEngine_ConnectNotify(self QAudioEngine, cb intptr_t, signal *QMetaMethod) {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(signal *qt.QMetaMethod), signal *qt.QMetaMethod))
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	slotval1 := qt.UnsafeNewQMetaMethod(unsafe.Pointer(signal))

	gofunc((&QAudioEngine{h: self}).callVirtualBase_ConnectNotify, slotval1)

}

func (this *QAudioEngine) callVirtualBase_DisconnectNotify(signal *qt.QMetaMethod) {

	QAudioEngine_virtualbase_DisconnectNotify(unsafe.Pointer(this.h), (*QMetaMethod)(signal.UnsafePointer()))

}
func (this *QAudioEngine) OnDisconnectNotify(slot func(super func(signal *qt.QMetaMethod), signal *qt.QMetaMethod)) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QAudioEngine_override_virtual_DisconnectNotify(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QAudioEngine_DisconnectNotify
func miqt_exec_callback_QAudioEngine_DisconnectNotify(self QAudioEngine, cb intptr_t, signal *QMetaMethod) {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(signal *qt.QMetaMethod), signal *qt.QMetaMethod))
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	slotval1 := qt.UnsafeNewQMetaMethod(unsafe.Pointer(signal))

	gofunc((&QAudioEngine{h: self}).callVirtualBase_DisconnectNotify, slotval1)

}
